provisioning/baremetal_secrets.go

```diff
@@ -0,0 +1,131 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+const (
+	baremetalSecretName = "metal3-mariadb-password"
+	baremetalSecretKey  = "password"
+	ironicUsernameKey   = "username"
+	ironicPasswordKey   = "password"
+	ironicHtpasswdKey   = "htpasswd"
+	ironicConfigKey     = "auth-config"
+	ironicSecretName    = "metal3-ironic-password"
+	ironicUsername      = "ironic-user"
+	inspectorSecretName = "metal3-ironic-inspector-password"
+	inspectorUsername   = "inspector-user"
+)
+
+func generateRandomPassword() (string, error) {
+	chars := []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz" +
+		"0123456789")
+	length := 16
+	buf := make([]rune, length)
+	numChars := big.NewInt(int64(len(chars)))
+	for i := range buf {
+		c, err := rand.Int(rand.Reader, numChars)
+		if err != nil {
+			return "", err
+		}
+		buf[i] = chars[c.Uint64()]
+	}
+	return string(buf), nil
+}
+
+func createMariadbPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string) error {
+	_, err := client.Secrets(targetNamespace).Get(context.Background(), baremetalSecretName, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Secret does not already exist. So, create one.
+	password, err := generateRandomPassword()
+	if err != nil {
+		return err
+	}
+	_, err = client.Secrets(targetNamespace).Create(
+		context.Background(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      baremetalSecretName,
+				Namespace: targetNamespace,
+			},
+			StringData: map[string]string{
+				baremetalSecretKey: password,
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	return err
+}
+
+func createIronicPasswordSecret(client coreclientv1.SecretsGetter, targetNamespace string, name string, username string, configSection string) error {
+	_, err := client.Secrets(targetNamespace).Get(context.Background(), name, metav1.GetOptions{})
+	if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// Secret does not already exist. So, create one.
+	password, err := generateRandomPassword()
+	if err != nil {
+		return err
+	}
+	hash, err := bcrypt.GenerateFromPassword([]byte(password), 5) // Use same cost as htpasswd default
+	if err != nil {
+		return err
+	}
+	// Change hash version from $2a$ to $2y$, as generated by htpasswd.
+	// These are equivalent for our purposes.
+	hash[2] = 'y'
+
+	_, err = client.Secrets(targetNamespace).Create(
+		context.Background(),
+		&corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: targetNamespace,
+			},
+			StringData: map[string]string{
+				ironicUsernameKey: username,
+				ironicPasswordKey: password,
+				ironicHtpasswdKey: fmt.Sprintf("%s:%s", username, hash),
+				ironicConfigKey: fmt.Sprintf(`[%s]
+auth_type = http_basic
+username = %s
+password = %s
+`,
+					configSection, username, password),
+			},
+		},
+		metav1.CreateOptions{},
+	)
+	log.V(1).Info("created password for", "name", name)
```

Comment:
Can we combine the first two strings into one? And also shouldn't we be using Infof?

Reply:
Referring to https://github.com/openshift/cluster-baremetal-operator/blob/master/controllers/provisioning_controller.go#L71 and https://github.com/go-logr/logr/blob/master/examples/usage_example.go, this seems like it is OK.